### PR TITLE
Extract the logic to identify bug types into individual feature classes

### PR DESCRIPTION
**Pull request by PromiseFru**
Resolves: #3900

### CODE REVIEW

**Review by suhaibmujahid — COMMENTED** [review]
Thank you, @PromiseFru! Could you please modify the PR description to mention "Address part of:" instead of "Fixes:"? The scope of the issue is not limited to the performance part. Anyway, we can limit this PR to do the performance part for now.

File: bugbug/utils.py [hunk 1]
```diff
@@ -558,3 +558,54 @@ def escape_markdown(text: str) -> str:
...
+        "sec-": "security",
+        "csectype-": "security",
+        "memory-": "memory",
+        "crash": "crash",
+        "crashreportid": "crash",
+        "perf": "performance",
+        "topperf": "performance",
+        "main-thread-io": "performance",
+        "power": "power",
+    }
```
[the remark this answers is not in the document]
Comment (suhaibmujahid) [reply bugbug/utils.py]: Modifying this PR should be fine since we are still within the scope of the issue. At that point, `bug_to_types` itself should be moved as well. Since we are on it, I would rename the function to something more descriptive like `identify_bug_types` or `infer_bug_types`.

File: bugbug/models/bugtype.py [hunk 2]
```diff
@@ -103,7 +33,19 @@ def __init__(self, lemmatization=False, historical=False):
...
+                    "sec-",
+                    "csectype-",
+                    "memory-",
+                    "crash",
+                    "crashreportid",
+                    "perf",
+                    "topperf",
+                    "main-thread-io",
+                    "power",
+                }
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:47]: I feel this is a bit fragile and could be an error-prone. If any of the functions add a new keyword, it will be highly likely to miss adding it here.

One solution would be to convert the functions to feature classes. Then, the keywords will be stored as attributes from these classes:

```py
self.label_extractors = [
    IsPerformanceBug(),
    IsMemoryBug(),
    IsPowerBug(),
    IsSecurityBug(),
    IsCrashBug(),
]
```

We could extract the keywords by iterating over the label extractors:

```py
keywords = {
    keyword
    for extractor in self.label_extractors
    for keyword in extractor.keywords
}
```

Comment (PromiseFru) [reply bugbug/models/bugtype.py]: Updated:

https://github.com/mozilla/bugbug/blob/906aa2bb12944ffa1f02a7837cb9f3fb0b9cc46a/bugbug/models/bugtype.py#L30-L51

Comment (suhaibmujahid) [reply bugbug/models/bugtype.py]: Nice! Thank you.

File: bugbug/models/bugtype.py [hunk 3]
```diff
@@ -158,7 +102,7 @@ def __init__(self, lemmatization=False, historical=False):
                 (
                     "estimator",
                     OneVsRestClassifier(
-                        xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+                        xgboost.XGBClassifier(n_jobs=get_physical_cpu_count())
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:105]: Let we try avoid unrelated changes

File: bugbug/models/bugtype.py [hunk 4]
```diff
@@ -98,12 +28,26 @@ def __init__(self, lemmatization=False, historical=False):
...
 
+        self.label_extractors = [
+            bug_features.IsPerformanceBug(),
+            bug_features.IsMemoryBug(),
+            bug_features.IsPowerBug(),
+            bug_features.IsSecurityBug(),
+            bug_features.IsCrashBug(),
+        ]
+
+        keywords = {
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:39]: I would use a more descriptive name (e.g., `label_keywords`).

File: bugbug/models/bugtype.py [hunk 5]
```diff
@@ -171,7 +115,7 @@ def get_labels(self) -> tuple[dict[int, np.ndarray], list[str]]:
 
         for bug_data in bug_map.values():
             target = np.zeros(len(TYPE_LIST))
-            for type_ in bug_to_types(bug_data, bug_map):
+            for type_ in bug_features.infer_bug_types(bug_data, bug_map):
                 target[TYPE_LIST.index(type_)] = 1
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:119]: We could use `self.label_extractors` here instead:
```suggestion
            for i, is_type in enumerate(self.label_extractors):
                if is_type(bug_data, bug_map)
                    target[i] = 1
```

In this way, we will not need to use `bug_features.infer_bug_types()` at all. Also, we will not need `TYPE_LIST ` as well.

File: bugbug/bug_features.py [hunk 6]
```diff
@@ -690,3 +691,185 @@ def apply_rollback(bugs_iter):
...
+            keyword.startswith(keyword_start)
+            for keyword_start in self.keywords
+            for keyword in bug["keywords"]
+        ):
+            return True
+
+        return False
+
+
+def infer_bug_types(
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:838]: We could convert it to a class and name it `BugTypes`.

File: bugbug/bug_features.py [hunk 7]
```diff
@@ -690,3 +691,185 @@ def apply_rollback(bugs_iter):
...
+    - list[str]: A list of inferred bug types (e.g., "memory", "power",
+        "performance", "security", "crash").
+    """
+    is_performance_bug = IsPerformanceBug()
+    is_memory_bug = IsMemoryBug()
+    is_power_bug = IsPowerBug()
+    is_security_bug = IsSecurityBug()
+    is_crash_bug = IsCrashBug()
+
+    types = set()
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:858]: We do not need a set here. We do not need to deal with duplications here.

File: bugbug/bug_features.py [hunk 8]
```diff
@@ -690,3 +691,185 @@ def apply_rollback(bugs_iter):
...
+    if is_performance_bug(bug):
+        types.add("performance")
+
+    if is_security_bug(bug):
+        types.add("security")
+
+    if is_crash_bug(bug):
+        types.add("crash")
+
+    return list(types)
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:875]: I would store the functions in a list and loop over it to extract the types.
```suggestion
    return [
        is_type.type_name
        for is_type in self.bug_type_extractors
        if is_type(bug_data)
    ]
```

File: bugbug/models/bugtype.py (hunk 5 above)
Comment (PromiseFru) [reply bugbug/models/bugtype.py]: Since the `bug_type_extractors` list was created in the `BugTypes` class:

https://github.com/mozilla/bugbug/blob/e5b0b2f52a8774e9a524bc97123f7afd3b9e018e/bugbug/bug_features.py#L853-L863

It can be used to generate the `label_keywords` set:

https://github.com/mozilla/bugbug/blob/e5b0b2f52a8774e9a524bc97123f7afd3b9e018e/bugbug/models/bugtype.py#L27-L36

File: bugbug/bug_features.py [hunk 9]
```diff
@@ -690,3 +691,194 @@ def apply_rollback(bugs_iter):
...
+    type_name = "performance"
+    keywords = {"perf", "topperf", "main-thread-io"}
+
+    def __call__(
+        self,
+        bug: bugzilla.BugDict,
+        bug_map: Optional[dict[int, bugzilla.BugDict]] = None,
+    ) -> bool:
+        if any(
+            f"[{whiteboard_text}" in bug["whiteboard"].lower()
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:709]: Instead of adding `[` at the run time, I would have it as part of the prefixes in the list.

File: bugbug/bug_features.py [hunk 10]
```diff
@@ -690,3 +691,194 @@ def apply_rollback(bugs_iter):
...
+            for whiteboard_text in (
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:710]: I would change it to be `prefix`, similar to the keywords:
```suggestion
            for prefix in (
```

File: bugbug/bug_features.py [hunk 11]
```diff
@@ -690,3 +691,194 @@ def apply_rollback(bugs_iter):
...
+    """Determine if the bug is related to performance based on given bug data."""
+
+    name = "Is Performance Bug"
+    type_name = "performance"
+    keywords = {"perf", "topperf", "main-thread-io"}
+
+    def __call__(
+        self,
+        bug: bugzilla.BugDict,
+        bug_map: Optional[dict[int, bugzilla.BugDict]] = None,
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:706]: See https://github.com/mozilla/bugbug/issues/3856

**Review by suhaibmujahid — CHANGES_REQUESTED** [review]

File: bugbug/bug_features.py [hunk 12]
```diff
@@ -690,3 +691,194 @@ def apply_rollback(bugs_iter):
...
 
         return pd.DataFrame(apply_transform(bug) for bug in bugs_iter)
+
+
+class IsPerformanceBug(SingleBugFeature):
+    """Determine if the bug is related to performance based on given bug data."""
+
+    name = "Is Performance Bug"
+    type_name = "performance"
+    keywords = {"perf", "topperf", "main-thread-io"}
```
[the remark this answers is not in the document]
Comment (PromiseFru) [reply bugbug/bug_features.py]: I used the set as an extra safety measure against duplicates, but I don't think anyone would accidentally input duplicates unless the entries are quite extensive.

**Comment by PromiseFru** [comment]
@suhaibmujahid: I noticed that `dateutil.parser` was imported twice. I don't think it's necessary, is it?

https://github.com/mozilla/bugbug/blob/8e4235c7f7b765247e2b305d62d763d293095243/bugbug/bug_features.py#L13-L15

File: bugbug/models/bugtype.py [hunk 13]
```diff
@@ -18,92 +18,28 @@
...
+            for keyword in extractor.keyword_prefixes
+        }
+
         feature_extractors = [
             bug_features.HasSTR(),
             bug_features.Severity(),
             # Ignore keywords that would make the ML completely skewed
             # (we are going to use them as 100% rules in the evaluation phase).
-            bug_features.Keywords(set(KEYWORD_DICT.keys())),
+            bug_features.Keywords(label_keywords),
```
[the remark this answers is not in the document]
Comment (PromiseFru) [reply bugbug/models/bugtype.py:42]: > The `Keywords` feature accepts keywords but is given keyword prefixes here. We could add a new parameter to accept prefixes.

@suhaibmujahid, I was wondering, would it be best to add a `prefix` bool parameter to treat the keywords as prefixes or include a `prefixes_to_ignore` set that coexists with the `to_ignore` set parameters?

[the remark this answers is not in the document]
Comment (suhaibmujahid) [reply bugbug/models/bugtype.py:42]: I think including `prefixes_to_ignore` would be a better option.

[the remark this answers is not in the document]
Comment (suhaibmujahid) [reply bugbug/models/bugtype.py:42]: Since it is a pre-existing bug, let us fix it in a separate PR.

[the remark this answers is not in the document]
Comment (suhaibmujahid) [reply bugbug/models/bugtype.py:42]: I filed an issue: https://github.com/mozilla/bugbug/issues/3926

File: bugbug/models/bugtype.py [hunk 14]
```diff
@@ -18,92 +18,28 @@
...
-
 
 class BugTypeModel(BugModel):
     def __init__(self, lemmatization=False, historical=False):
         BugModel.__init__(self, lemmatization)
 
         self.calculate_importance = False
 
+        bug_types = bug_features.BugTypes()
+        self.label_extractors = bug_types.bug_type_extractors
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:29]: We could simplify this:
```suggestion
        self.label_extractors = BugTypes.bug_type_extractors
```

File: bugbug/models/bugtype.py [hunk 15]
```diff
@@ -18,92 +18,28 @@
...
+
+        label_keywords = {
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:31]: Using `prefixes` would be more descriptive:
```suggestion
        label_keyword_prefixes = {
```

File: bugbug/models/bugtype.py [hunk 16]
```diff
@@ -170,20 +106,23 @@ def get_labels(self) -> tuple[dict[int, np.ndarray], list[str]]:
...
-                target[TYPE_LIST.index(type_)] = 1
+            target = np.zeros(len(self.label_extractors))
+            for i, is_type in enumerate(self.label_extractors):
+                if is_type(bug_data, bug_map):
+                    target[i] = 1
 
             classes[int(bug_data["id"])] = target
 
-        for type_ in TYPE_LIST:
+        type_list = [extractor.type_name for extractor in self.label_extractors]
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:116]: More descriptive and consistent with `bug_type` that we use in the loop:
```suggestion
        bug_types = [extractor.type_name for extractor in self.label_extractors]
```

File: bugbug/models/bugtype.py [hunk 17]
```diff
@@ -194,11 +133,14 @@ def overwrite_classes(
...
     ):
+        bug_map = {bug["id"]: bug for bug in bugs}
+
         for i, bug in enumerate(bugs):
-            for type_ in bug_to_types(bug):
-                if probabilities:
-                    classes[i][TYPE_LIST.index(type_)] = 1.0
-                else:
-                    classes[i][TYPE_LIST.index(type_)] = 1
+            for j, is_type in enumerate(self.label_extractors):
```
Comment (suhaibmujahid) [inline bugbug/models/bugtype.py:139]: I know that I suggested the old name, but I fell these name is more descriptive:
```suggestion
            for j, is_type_applicable in enumerate(self.bug_type_extractors):
```

**Review by suhaibmujahid — COMMENTED** [review]
Thank you, @PromiseFru! It look good, I just have some variable name suggestions to improve the readability.

File: bugbug/bug_features.py [hunk 18]
```diff
@@ -690,3 +690,196 @@ def apply_rollback(bugs_iter):
...
+    type_name = "power"
+    keyword_prefixes = ("power",)
+    whiteboard_prefixes = ("[power",)
+
+    def __call__(
+        self,
+        bug: bugzilla.BugDict,
+        bug_map: dict[int, bugzilla.BugDict] | None = None,
+    ) -> bool:
+        bug_whiteboard = bug["whiteboard"].lower()
```
Comment (suhaibmujahid) [inline bugbug/bug_features.py:783]: We may want to move this to right before where we use it.

**Review by suhaibmujahid — DISMISSED** [review]
Thank you! LGTM!

Could you please check the model accuracy metrics before and after the refactoring? They should be in the same range.

**Comment by PromiseFru** [comment]
> Thank you! LGTM!
> 
> Could you please check the model accuracy metrics before and after the refactoring? They should be in the same range.

I cannot obtain the performance metrics similar to the situation encountered [here](https://github.com/mozilla/bugbug/pull/3823#issuecomment-1824888388).

<details>
    <summary>Training before refactoring</summary>

    2023-12-16 10:58:18,674:INFO:__main__:Training *bugtype* model
    2023-12-16 10:58:41,565:INFO:bugbug.models.bugtype:7544 crash bugs
    2023-12-16 10:58:41,629:INFO:bugbug.models.bugtype:560 memory bugs
    2023-12-16 10:58:41,696:INFO:bugbug.models.bugtype:4730 performance bugs
    2023-12-16 10:58:41,766:INFO:bugbug.models.bugtype:124 power bugs
    2023-12-16 10:58:41,839:INFO:bugbug.models.bugtype:2575 security bugs
    Traceback (most recent call last):
    File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/runpy.py", line 196, in _run_module_as_main
        return _run_code(code, main_globals, None,
    File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/runpy.py", line 86, in _run_code
        exec(code, run_globals)
    File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 145, in <module>
        main()
    File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 141, in main
        retriever.go(args)
    File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 41, in go
        metrics = model_obj.train(limit=args.limit)
    File "/home/promisefru/mozilla/bugbug/bugbug/model.py", line 377, in train
        self.le.fit(y)
    File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/site-packages/sklearn/preprocessing/_label.py", line 98, in fit
        y = column_or_1d(y, warn=True)
    File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/site-packages/sklearn/utils/validation.py", line 1156, in column_or_1d
        raise ValueError(
    ValueError: y should be a 1d array, got an array of shape (198578, 5) instead.
</details>

<details>
    <summary>Training after refactoring</summary>

    2023-12-16 11:24:24,063:INFO:__main__:Training *bugtype* model
    2023-12-16 11:24:47,247:INFO:bugbug.models.bugtype:7521 crash bugs
    2023-12-16 11:24:47,293:INFO:bugbug.models.bugtype:559 memory bugs
    2023-12-16 11:24:47,339:INFO:bugbug.models.bugtype:4612 performance bugs
    2023-12-16 11:24:47,385:INFO:bugbug.models.bugtype:124 power bugs
    2023-12-16 11:24:47,432:INFO:bugbug.models.bugtype:2575 security bugs
    Traceback (most recent call last):
      File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/runpy.py", line 196, in _run_module_as_main
        return _run_code(code, main_globals, None,
      File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/runpy.py", line 86, in _run_code
        exec(code, run_globals)
      File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 145, in <module>
        main()
      File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 141, in main
        retriever.go(args)
      File "/home/promisefru/mozilla/bugbug/scripts/trainer.py", line 41, in go
        metrics = model_obj.train(limit=args.limit)
      File "/home/promisefru/mozilla/bugbug/bugbug/model.py", line 377, in train
        self.le.fit(y)
      File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/site-packages/sklearn/preprocessing/_label.py", line 98, in fit
        y = column_or_1d(y, warn=True)
      File "/home/promisefru/anaconda3/envs/bugbug/lib/python3.10/site-packages/sklearn/utils/validation.py", line 1156, in column_or_1d
        raise ValueError(
    ValueError: y should be a 1d array, got an array of shape (198578, 5) instead.
</details>

File: tests/test_bug_features.py [hunk 19]
```diff
@@ -178,3 +185,70 @@ def test_BugExtractor():
...
+def test_bug_types() -> None:
+    bug_types = BugTypes()
+
+    bug_map = {int(bug["id"]): bug for bug in bugzilla.get_bugs(include_invalid=True)}
+
+    assert bug_types(bug_map[447581]) == ["performance"]
+    assert bug_types(bug_map[1325215], bug_map) == ["memory"]
+    assert bug_types(bug_map[922874]) == ["power"]
+    assert bug_types(bug_map[528988]) == ["security"]
+    assert bug_types(bug_map[1046231]) == ["crash"]
```
Comment (suhaibmujahid) [inline tests/test_bug_features.py:254]: We could simplify this by following the same patterns that we use in the rest of the test file. There is no need to test the individual feature classes since they are already used in `BugTypes()`. Thus, instead of adding bugs to the `tests/fixtures/bugs.json` file, you could create a new file for this (e.g., `tests/fixtures/bug_features/bug_typs.json`).

**Review by suhaibmujahid — APPROVED** [review]
